Lift fitting of costs: From within algorithm implementation, to ChangeDetector

johannvk opened the pull request:
To separate concerns this PR lifts the fitting of the costs and related objects used by `ChangeDetector`s from the respective methods implementing the change detection algorithms, up to the internal `ChangeDetector._predict(self, X)` methods.

This hopefully makes the dependencies between different components of the change detection pipeline more clearly visible.

Tveten reviewed (commented):
Two things that should be fixed:

- MovingWindow should also have its scorer fitted before being sent to `moving_window_transform`.
- The `num_observations` argument: This can be obtained more safely from the fitted scorer objects by `scorer._X.shape[0]`. No need as a separate argument.

   - I'm using the convention that `n` are the number of rows in the data and `p` the number of columns throughout the package. So for consistency I don't see the need for changing this here to either `num_observations` or `num_obs`. But not a blocker for merging :)

johannvk commented:
> Two things that should be fixed:
> 
>     * MovingWindow should also have its scorer fitted before being sent to `moving_window_transform`.
> 
>     * The `num_observations` argument: This can be obtained more safely from the fitted scorer objects by `scorer._X.shape[0]`. No need as a separate argument.
>       
>       * I'm using the convention that `n` are the number of rows in the data and `p` the number of columns throughout the package. So for consistency I don't see the need for changing this here to either `num_observations` or `num_obs`. But not a blocker for merging :)

I applied the "lifting" to the `MovingWindow` change detector as well now, and renamed `num_obs[ervations]` to `n_samples` following `skchange` convention. Also switched to retrieving `n_samples` from the fitted `BaseIntervalScore`-objects as you mentioned, instead of passing in to the change detector algorithms as an additional parameter.
